wpsc-admin/ajax.php

```diff
@@ -456,6 +456,189 @@ function _wpsc_ajax_purchase_log_action_link() {
 }
 
 /**
+ * Remove purchase log item.
+ *
+ * @since   4.0
+ * @access  private
+ *
+ * @return  array|WP_Error  $return  Response args if successful, WP_Error if otherwise
+ */
+function _wpsc_ajax_remove_log_item() {
+
+	if ( isset( $_POST['item_id'], $_POST['log_id'] ) ) {
+
+		$item_id = absint( $_POST['item_id'] );
+		$log_id  = absint( $_POST['log_id'] );
+		$log     = new WPSC_Purchase_Log( $log_id );
+
+
+		if ( $log->remove_cart_item( $item_id ) ) {
+			return _wpsc_init_log_items( $log );
+		}
+	}
+
+	return new WP_Error( 'wpsc_ajax_invalid_remove_log_item', __( 'Removing log item failed.', 'wp-e-commerce' ) );
+}
+
+/**
+ * Update purchase log item quantity.
+ *
+ * @since   4.0
+ * @access  private
+ *
+ * @return  array|WP_Error  $return  Response args if successful, WP_Error if otherwise
+ */
+function _wpsc_ajax_update_log_item_qty() {
+
+	if ( isset( $_POST['item_id'], $_POST['log_id'], $_POST['qty'] ) ) {
+
+		if ( empty( $_POST['qty'] ) ) {
+			return _wpsc_ajax_remove_log_item();
+		}
+
+		$item_id = absint( $_POST['item_id'] );
+		$log_id  = absint( $_POST['log_id'] );
+		$log     = new WPSC_Purchase_Log( $log_id );
+		$result  = $log->update_cart_item( $item_id, array( 'quantity' => absint( $_POST['qty'] ) ) );
+
+		if ( 0 === $result ) {
+			return true;
+		} elseif ( false !== $result ) {
+			return _wpsc_init_log_items( $log );
+		}
+	}
+
+	return new WP_Error( 'wpsc_ajax_invalid_update_log_item_qty', __( 'Updating log item quantity failed.', 'wp-e-commerce' ) );
+}
+
+/**
+ * Add purchase log item.
+ *
+ * @since   4.0
+ * @access  private
+ *
+ * @return  array|WP_Error  $return  Response args if successful, WP_Error if otherwise
+ */
+function _wpsc_ajax_add_log_item() {
+	global $wpsc_cart;
+
+	if ( isset( $_POST['product_ids'], $_POST['log_id'] ) && is_array( $_POST['product_ids'] ) ) {
+
+		$item_ids = array();
+
+		foreach ( $_POST['product_ids'] as $product_id ) {
+			$product_id = absint( $product_id );
+			$log_id  = absint( $_POST['log_id'] );
+			$log     = new WPSC_Purchase_Log( $log_id );
+
+			$item = new wpsc_cart_item( $product_id, array(), $wpsc_cart );
+			$item_id = $item->save_to_db( $log_id );
+			$item_ids[] = absint( $item_id );
+		}
+
+		return _wpsc_init_log_items( $log, $item_ids );
```

Comment:
If $_POST['product_ids'] is set to an empty array(), it will pass the first condition, but not be iterated through, and the $log variable will not exist.